README.md
PlanetaryPy
================

<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->

Potential
logo:![image.png](images/epilup_with_python_logo_with_axis.png)

## Install

``` bash
pip install planetarypy
```

This will pull in these other dependencies and their dependencies:

`tomlkit pandas pvl numpy python-dateutil tqdm lxml yarl hirise-tools kalasiris`

## Suggested standard abbreviations:

- Inside these docs the package will be called `PLPY` for brevity.
- A standard Python import could be: `plp` or `plpy`
  - because the last `p` in `plp` can be pronounced out, we consider
    these equivalent for human conversation and pronounce these
    “plippy”.

## General scope

First and foremost this package should provide support in working with
planetary science data.

With `working` we mean:

- locating
- retrieving
- reading
- further processing

of data.

### Locating

This library manages, via its `PDS tools`, multiple PDS3 index files per
instrument that can be used for identifying data of interest. These
index files are automatically downloaded and converted to the very
performant (and cloud-ready)
[parquet](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_parquet.html)
file format. \> Parquet is able to store advanced datatypes like
nan-capable integer and full datetime objects, as opposed to HDF5.

### Retrieving

The interface to getting data is via a path-retrieving function based on
a PDS product-id. If that product-id is available locally, the path will
be returned. If it is not, it will previously be downloaded, stored in a
systematic fashion organized by mission and instrument, and then the
local path will be returned.

### Reading

For now, the library only returns the path to the object and the user
needs to sort out the reading process. A recently funded NASA project
`Planetary Data Reader` will be integrated here, so that basic reading
into memory can be provided.

As such, we anticipate two classes of reading support: 1. basic reading
into numpy and/or xarray 1. added reader functionality like basic plots
and basic geospatial processing, as supported by interested parties

There will exist larger other packages that focus on working with a
given instrument’s data, in which case that package could become an
affiliated package with the `planetarypy` GitHub organization, if so
desired.

### Further processing

In the future, additional frequently used procedures will be added to
this library, e.g. \* frequently used GDAL/rasterio procedures \*
frequently used SPICE operations \* like surface illumination on a given
body

## PDS tools

Look at the `Apps` docs to see what `pds.apps` exist for easily getting
PDS indexes. The `find_index` app is specifically useful when you don’t
know what index files exist.

So far, the following indexes are supported (but not necessarily all
tested within PLPY):

- Cassini
  - ISS (all)
  - UVIS (all)
- MRO
  - CTX EDR
  - HiRISE
    - EDR, RDR, DTM
      - EDR index has a bug (as delivered by the team, reported), where
        I need to activate an existing fix for it.
- LRO
  - Diviner (DLRE)
    - EDR, RDR
  - LOLA
    - EDR, RDR

### More indexes

More indexes of other instruments can be easily added by following the
existing structure of what has been copied into your config at
`~/.planetarypy_config.toml`.

Please consider submitting a pull request for adding further PDS index
files into the config file at its source:
https://github.com/michaelaye/nbplanetary/blob/master/planetarypy/data/planetarypy_config.toml

## Utils

Find something in `Utils` for working with NASA timestamps and a well
working URL download function
[`url_retrieve`](https://michaelaye.github.io/planetarypy/api/utils.html#url_retrieve),
among other stuff.

## Experiment/Instrument Specific

So far, `planetarypy` supports CTX EDR and HiRISE RGB.NOMAP data. Look
at the
[`CTX`](https://michaelaye.github.io/planetarypy/api/ctx.html#ctx) and
`HiRISE` pages for descriptions of classes for working with these data.

## Bug reports

Please submit bug reports at
https://github.com/michaelaye/nbplanetary/issues

## How to use

### Indexes

See [PDS apps](api/02c_pds.apps.ipynb) for more details.

``` python
from planetarypy.pds.apps import get_index
```

``` python
ctrindex = get_index("mro.ctx", "edr")
ctrindex.sample(5, random_state=42)  # setting random_state to always get same files for docs
```

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>VOLUME_ID</th>
      <th>FILE_SPECIFICATION_NAME</th>
      <th>ORIGINAL_PRODUCT_ID</th>
      <th>PRODUCT_ID</th>
      <th>IMAGE_TIME</th>
      <th>INSTRUMENT_ID</th>
      <th>INSTRUMENT_MODE_ID</th>
      <th>LINE_SAMPLES</th>
      <th>LINES</th>
      <th>SPATIAL_SUMMING</th>
      <th>...</th>
      <th>SUB_SOLAR_LATITUDE</th>
      <th>SUB_SPACECRAFT_LONGITUDE</th>
      <th>SUB_SPACECRAFT_LATITUDE</th>
      <th>SOLAR_DISTANCE</th>
      <th>SOLAR_LONGITUDE</th>
      <th>LOCAL_TIME</th>
      <th>IMAGE_SKEW_ANGLE</th>
      <th>RATIONALE_DESC</th>
      <th>DATA_QUALITY_DESC</th>
      <th>ORBIT_NUMBER</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>102927</th>
      <td>MROX_3290</td>
      <td>DATA/K09_056855_1472_XN_32S152W.IMG</td>
      <td>4A_04_112001B600</td>
      <td>K09_056855_1472_XN_32S152W</td>
      <td>2018-09-12 15:27:23.819</td>
      <td>CTX</td>
      <td>NIFL</td>
      <td>5056</td>
      <td>7168</td>
      <td>1</td>
      <td>...</td>
      <td>-23.61</td>
      <td>152.16</td>
      <td>-32.84</td>
      <td>206674590.7</td>
      <td>248.70</td>
      <td>15.12</td>
      <td>90.1</td>
      <td>Ride-along with HiRISE</td>
      <td>OK</td>
      <td>56855</td>
    </tr>
    <tr>
      <th>116952</th>
      <td>MROX_3714</td>
      <td>DATA/N08_065442_0965_XN_83S191W.IMG</td>
      <td>4A_04_114C019800</td>
      <td>N08_065442_0965_XN_83S191W</td>
      <td>2020-07-12 17:31:23.962</td>
      <td>CTX</td>
      <td>NIFL</td>
      <td>5056</td>
      <td>33792</td>
      <td>1</td>
      <td>...</td>
      <td>-21.24</td>
      <td>190.23</td>
      <td>-83.51</td>
      <td>207154930.4</td>
      <td>237.37</td>
      <td>17.12</td>
      <td>91.4</td>
      <td>Planum Australe</td>
      <td>OK</td>
      <td>65442</td>
    </tr>
    <tr>
      <th>16386</th>
      <td>MROX_0690</td>
      <td>DATA/B01_010083_2567_XI_76N181W.IMG</td>
      <td>4A_04_103101E100</td>
      <td>B01_010083_2567_XI_76N181W</td>
      <td>2008-09-20 04:33:05.584</td>
      <td>CTX</td>
      <td>ITL</td>
      <td>5056</td>
      <td>52224</td>
      <td>1</td>
      <td>...</td>
      <td>19.30</td>
      <td>180.80</td>
      <td>76.84</td>
      <td>237461101.1</td>
      <td>129.80</td>
      <td>14.76</td>
      <td>89.7</td>
      <td>North polar outlier</td>
      <td>OK</td>
      <td>10083</td>
    </tr>
    <tr>
      <th>107515</th>
      <td>MROX_3450</td>
      <td>DATA/K15_059059_1434_XN_36S196W.IMG</td>
      <td>4A_04_112C004300</td>
      <td>K15_059059_1434_XN_36S196W</td>
      <td>2019-03-03 08:44:14.027</td>
      <td>CTX</td>
      <td>NIFL</td>
      <td>2528</td>
      <td>9216</td>
      <td>1</td>
      <td>...</td>
      <td>-4.38</td>
      <td>196.66</td>
      <td>-36.69</td>
      <td>229158765.3</td>
      <td>349.79</td>
      <td>14.21</td>
      <td>90.1</td>
      <td>Craters in Terra Cimmeria</td>
      <td>OK</td>
      <td>59059</td>
    </tr>
    <tr>
      <th>57288</th>
      <td>MROX_1890</td>
      <td>DATA/D16_033513_1653_XI_14S108W.IMG</td>
      <td>4A_04_10A9016F00</td>
      <td>D16_033513_1653_XI_14S108W</td>
      <td>2013-09-19 20:45:34.363</td>
      <td>CTX</td>
      <td>ITL</td>
      <td>5056</td>
      <td>15360</td>
      <td>1</td>
      <td>...</td>
      <td>10.17</td>
      <td>108.53</td>
      <td>-14.83</td>
      <td>241359283.4</td>
      <td>24.21</td>
      <td>14.66</td>
      <td>90.2</td>
      <td>Syria/Claritas dust-raising event monitoring site</td>
      <td>OK</td>
      <td>33513</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 51 columns</p>
</div>

``` python
hirise_rdr = get_index("mro.hirise", "rdr")
hirise_rdr.sample(5, random_state=42)
```

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>VOLUME_ID</th>
      <th>FILE_NAME_SPECIFICATION</th>
      <th>INSTRUMENT_HOST_ID</th>
      <th>INSTRUMENT_ID</th>
      <th>OBSERVATION_ID</th>
      <th>PRODUCT_ID</th>
      <th>PRODUCT_VERSION_ID</th>
      <th>TARGET_NAME</th>
      <th>ORBIT_NUMBER</th>
      <th>MISSION_PHASE_NAME</th>
      <th>...</th>
      <th>LINE_PROJECTION_OFFSET</th>
      <th>SAMPLE_PROJECTION_OFFSET</th>
      <th>CORNER1_LATITUDE</th>
      <th>CORNER1_LONGITUDE</th>
      <th>CORNER2_LATITUDE</th>
      <th>CORNER2_LONGITUDE</th>
      <th>CORNER3_LATITUDE</th>
      <th>CORNER3_LONGITUDE</th>
      <th>CORNER4_LATITUDE</th>
      <th>CORNER4_LONGITUDE</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>95782</th>
      <td>MROHR_0001</td>
      <td>RDR/ESP/ORB_049400_049499/ESP_049456_0945/ESP_...</td>
      <td>MRO</td>
      <td>HIRISE</td>
      <td>ESP_049456_0945</td>
      <td>ESP_049456_0945_COLOR</td>
      <td>1</td>
      <td>MARS</td>
      <td>49456</td>
      <td>Extended Science Phase</td>
      <td>...</td>
      <td>289555.5</td>
      <td>1014477.5</td>
      <td>-85.6701</td>
      <td>286.4690</td>
      <td>-85.6815</td>
      <td>286.3020</td>
      <td>-85.5562</td>
      <td>284.8560</td>
      <td>-85.5452</td>
      <td>285.0210</td>
    </tr>
    <tr>
      <th>142902</th>
      <td>MROHR_0001</td>
      <td>RDR/ESP/ORB_072700_072799/ESP_072755_2030/ESP_...</td>
      <td>MRO</td>
      <td>HIRISE</td>
      <td>ESP_072755_2030</td>
      <td>ESP_072755_2030_RED</td>
      <td>1</td>
      <td>MARS</td>
      <td>72755</td>
      <td>Extended Science Phase</td>
      <td>...</td>
      <td>5448057.5</td>
      <td>-36148404.0</td>
      <td>22.8511</td>
      <td>342.4710</td>
      <td>22.8408</td>
      <td>342.3770</td>
      <td>22.9836</td>
      <td>342.3590</td>
      <td>22.9940</td>
      <td>342.4520</td>
    </tr>
    <tr>
      <th>114241</th>
      <td>MROHR_0001</td>
      <td>RDR/ESP/ORB_057400_057499/ESP_057400_1150/ESP_...</td>
      <td>MRO</td>
      <td>HIRISE</td>
      <td>ESP_057400_1150</td>
      <td>ESP_057400_1150_RED</td>
      <td>1</td>
      <td>MARS</td>
      <td>57400</td>
      <td>Extended Science Phase</td>
      <td>...</td>
      <td>-15324106.0</td>
      <td>9551808.0</td>
      <td>-65.0631</td>
      <td>99.2951</td>
      <td>-65.0739</td>
      <td>99.1184</td>
      <td>-64.9297</td>
      <td>99.0696</td>
      <td>-64.9190</td>
      <td>99.2454</td>
    </tr>
    <tr>
      <th>260</th>
      <td>MROHR_0001</td>

CHANGELOG.md
# Release notes

<!-- do not remove -->

## 0.26.0

### New Features

- get Spicer running ([#37](https://github.com/michaelaye/nbplanetary/issues/37))
  - Spicer is my small class-based framework around SPICE commands to make 
illumination calculations more human-readable. I have used this for years and added it here now.﻿

- add self rep for CTXCollection ([#36](https://github.com/michaelaye/nbplanetary/issues/36))

- confusing/convoluted API for downloading indexes ([#35](https://github.com/michaelaye/nbplanetary/issues/35))
  - downloading was safeguarded far too much to be useful.
Removed a lot of code (always good), which also made `get_index` 
much clearer and simpler.﻿



## 0.25.0

### New Features

- Add generic SPICE kernel management ([#34](https://github.com/michaelaye/nbplanetary/issues/34))

- Add functionalities to CTXCollection ([#33](https://github.com/michaelaye/nbplanetary/issues/33))
  - CTXCollection should be able to use metadata from the index to:

    from the list of product_ids which is the essential attribute of a CTXCollection:

    - filter out corrupted data
    - filter for full width data of the sensor (important for flatfield calculations and mosaicking)
    - get the list of image times
    - get a dataframe of metadata filtered for the current list of product_ids.

- copy new config file to package data ([#32](https://github.com/michaelaye/nbplanetary/issues/32))

- Add dart to supported SPICE kernel archives ([#31](https://github.com/michaelaye/nbplanetary/issues/31))

- make storage more flexible ([#30](https://github.com/michaelaye/nbplanetary/issues/30))
  - sometimes there is a server with raw/edr data locally available.
But then processed data cannot be stored on that.
And sometimes this is true for one instrument or mission, but not the next.
add missions.<mission>.<instrument>.datapaths section to config file to manage this.
Also add DBManager class as general interface to all data storage things.

- add progress bar to fix_hirise_edr ([#25](https://github.com/michaelaye/nbplanetary/issues/25))

- add label downloader for HiRISE products ([#24](https://github.com/michaelaye/nbplanetary/issues/24))

- Run fix_hirise_edr automatically ([#23](https://github.com/michaelaye/nbplanetary/issues/23))
  - The `fix_hirise_edr` pds.index util exists but is not automatically called upon when
downloading the EDR index for MRO HiRISE

- Improve API for get_index() ([#22](https://github.com/michaelaye/nbplanetary/issues/22))
  - Index can read a shortened index key (without '.indexes.') so no need to make the user to know that.

- add read_index_data ([#21](https://github.com/michaelaye/nbplanetary/issues/21))

## 0.21.8

### New Features

- add stitched_cube_path to SOURCE_PRODUCT ([#20](https://github.com/michaelaye/nbplanetary/issues/20))



## 0.21.7


### Bugs Squashed

- url attribute was not working ([#19](https://github.com/michaelaye/nbplanetary/issues/19))


## 0.21.6

### New Features

- Consolidate classes better ([#18](https://github.com/michaelaye/nbplanetary/issues/18))
  - RGB_NOMAP had code that should be a base class
  - SOURCE_PRODUCT has some similarities to BASE_PRODUCT but not close enough to be derived
  - SOURCE_PRODUCT gets its own local_path/remote_path/url/download code segment
  - other products can inherit from SOURCE_PRODUCT and initialize it with the correct PRODUCT_ID string.﻿



## 0.21.5


### Bugs Squashed

- SOURCE_PRODUCT downloader was still old code ([#17](https://github.com/michaelaye/nbplanetary/issues/17))


## 0.21.4


### Bugs Squashed

- new product classes not exported to lib ([#16](https://github.com/michaelaye/nbplanetary/issues/16))


## 0.21.3

### New Features

- Add flag to not have URLs checked, for performance ([#15](https://github.com/michaelaye/nbplanetary/issues/15))

- Move check_url_exists to utils ([#13](https://github.com/michaelaye/nbplanetary/issues/13))

- Make sure get_index downloads and converts intelligently ([#12](https://github.com/michaelaye/nbplanetary/issues/12))

- Add SOURCE_PRODUCT, RED_PRODUCT, IR_PRODUCT to hirise module ([#11](https://github.com/michaelaye/nbplanetary/issues/11))

- Remove old HiRISE code ([#3](https://github.com/michaelaye/nbplanetary/issues/3))
  - The dependency on the old HiRISE package blocks creation of a conda package. 
  - Clean up old HiRISE code and add here.



## 0.21.2


### Bugs Squashed

- Missing dependency fastcore ([#10](https://github.com/michaelaye/nbplanetary/issues/10))


## 0.21.1


### Bugs Squashed

- User functions not exported to library ([#8](https://github.com/michaelaye/nbplanetary/issues/8))
  - I forgot the #export line in the cells with the user functions.



## 0.21.0

### New Features

- add user functions to spice.kernels ([#7](https://github.com/michaelaye/nbplanetary/issues/7))


## 0.20.0

### New Features

- Improve (date)time converters to be auto-matching ([#5](https://github.com/michaelaye/nbplanetary/issues/5))
  - Having to know the correct converter function to convert either just dates or full datetimes is annoying.

Improve the functionality so that only one function per direction is required:

* nasa_time_to_iso
* iso_time_to_nasa

- Add SPICE kernel retriever ([#4](https://github.com/michaelaye/nbplanetary/issues/4))
  - SPICE kernel management is an arduous task.

I have identified a way to wrap the NAIF Perl script for retrieving SPICE kernel lists and URLs for a given mission and start/stop times.

To be added within a new sub-module `spice.kernels`.





planetarypy/hirise.py
# AUTOGENERATED! DO NOT EDIT! File to edit: ../notebooks/api/04_hirise.ipynb.

# %% auto 0
__all__ = ['logger', 'storage_root', 'baseurl', 'rdrindex', 'OBSID', 'ProductPathfinder', 'COLOR_PRODUCT', 'RGB_NOMAP',
           'RGB_NOMAPCollection', 'SOURCE_PRODUCT', 'RED_PRODUCT', 'IR_PRODUCT', 'BG_PRODUCT']

# %% ../notebooks/api/04_hirise.ipynb 3
import logging
import warnings
import webbrowser

import rasterio
import rioxarray as rxr
from dask import compute, delayed
from fastcore.utils import Path
from yarl import URL

import hvplot.xarray  # noqa
from .config import config
from .pds.apps import get_index
from .utils import check_url_exists, url_retrieve

warnings.filterwarnings("ignore", category=rasterio.errors.NotGeoreferencedWarning)

# %% ../notebooks/api/04_hirise.ipynb 4
logger = logging.getLogger(__name__)

# %% ../notebooks/api/04_hirise.ipynb 5
storage_root = config.storage_root / "missions/mro/hirise"
baseurl = URL("https://hirise-pds.lpl.arizona.edu/PDS")
rdrindex = get_index("mro.hirise", "rdr")

# %% ../notebooks/api/04_hirise.ipynb 6
class OBSID:
    """Manage HiRISE observation ids.

    For example PSP_003092_0985.
    `phase` is set to PSP for orbits < 11000, no setting required.
    """

    def __init__(
        self,
        obsid: str,  # e.g. PSP_003092_0985
    ):
        phase, orbit, targetcode = obsid.split("_")
        self._orbit = int(orbit)
        self._targetcode = targetcode

    @property
    def orbit(self):
        return str(self._orbit).zfill(6)

    @orbit.setter
    def orbit(self, value: int):  # e.g. 11000, < 1_000_000
        if value > 999999:
            raise ValueError("Orbit cannot be larger than 999999")
        elif len(value) != 6:
            raise ValueError("Orbit string must be 6 digits.")
        self._orbit = value

    @property
    def targetcode(self):
        return self._targetcode

    @targetcode.setter
    def targetcode(
        self,
        value: str,  # e.g. "0985", must be 4 digits
    ):
        if len(str(value)) != 4:
            raise ValueError("Targetcode must be exactly 4 characters.")
        self._targetcode = value

    @property
    def phase(self):
        return "PSP" if int(self.orbit) < 11000 else "ESP"

    @property
    def id(self):
        return f"{self.phase}_{self.orbit}_{self.targetcode}"

    def __repr__(self):
        return self.__str__()

    def __str__(self):
        return self.id

    @property
    def upper_orbit_folder(self):
        """
        get the upper folder name wa
        rdrhere the given orbit folder is residing on the
        hisync server, e.g. 'ORB_011900_011999'
        """
        lower = int(self.orbit) // 100 * 100
        lowerstr = str(lower).zfill(6)
        upperstr = str(lower + 99).zfill(6)
        return f"ORB_{lowerstr}_{upperstr}"

    @property
    def storage_path_stem(self):
        return f"{self.phase}/{self.upper_orbit_folder}/{self.id}"

# %% ../notebooks/api/04_hirise.ipynb 13
class ProductPathfinder:
    """Determine paths and URLs for HiRISE RDR products (also EXTRAS.)

    We use the PDS definition of PRODUCT_ID here, e.g. PSP_003092_0985_RED.

    Attributes `jp2_path` and `label_path` get you the official RDR mosaic product,
    with `kind` steering if you get the COLOR or the RED product.
    All other properties go to the RDR/EXTRAS folder.
    The "PDS" part of the path is handled in the OBSID class.
    """

    kinds = ["RED", "BG", "IR", "COLOR", "IRB", "MIRB", "MRGB", "RGB", "RGB.NOMAP"]

    @classmethod
    def from_path(cls, path):
        path = Path(path)
        return cls(path.stem)

    def __init__(
        self,
        initstr: str,  # PRODUCT_ID string, e.g. PSP_003092_0985_RED
        check_url: bool = True,  # for performance, the user might not want the url check
    ):
        tokens = initstr.split("_")
        self._obsid = OBSID("_".join(tokens[:3]))
        try:
            self.kind = tokens[3]
        except IndexError:
            self._kind = None
        self.check_url = check_url

    @property
    def obsid(self):
        return self._obsid

    @obsid.setter
    def obsid(self, value: str):  # e.g. "PSP_003092_0985"
        self._obsid = OBSID(value)

    @property
    def kind(self):
        return self._kind

    @kind.setter
    def kind(
        self,
        value: str,  # one of "RED", "BG", "IR", "COLOR", "IRB", "MIRB", "MRGB", "RGB", "RGB.NOMAP"
    ):
        if value not in self.kinds:
            raise ValueError(f"kind must be in {self.kinds}")
        self._kind = value

    @property
    def product_id(self):
        return f"{self.obsid}_{self.kind}"

    def __str__(self) -> str:  # PRODUCT_ID string, e.g. PSP_003092_0985_RED
        return self.product_id

    def __repr__(self):
        return self.__str__()

    @property
    def storage_stem(
        self,
    ) -> str:  # e.g. 'PSP/ORB_003000_003099/PSP_003092_0985/PSP_003092_0985_RED'
        return f"{self.obsid.storage_path_stem}/{self.product_id}"

    @property
    def label_fname(self) -> str:  # e.g. 'PSP_003092_0985_RED.LBL'
        return f"{self.product_id}.LBL"

    @property
    def label_path(self):
        return Path("RDR") / (self.storage_stem + ".LBL")

    @property
    def local_label_path(self):
        path = self.label_path
        return storage_root / (f"{path.parent.name}/{path.name}")

    def download_label(self, overwrite=False):
        """Download the label file."""
        self.local_label_path.parent.mkdir(exist_ok=True, parents=True)
        if self.local_label_path.exists and not overwrite:
            return
        else:
            url_retrieve(self.label_url, self.local_label_path)

    def _make_url(self, obj):
        path = getattr(self, f"{obj}_path")
        url = baseurl / str(path)
        if self.check_url:
            if not check_url_exists(url):
                warnings.warn(f"{url} does not exist on the server.")
        return url

    def __getattr__(self, item):
        tokens = item.split("_")
        try:
            if tokens[-1] == "url":
                return self._make_url("_".join(tokens[:-1]))
        except IndexError:
            raise AttributeError(f"No attribute named '{item}' found.")

    @property
    def jp2_fname(self):
        return self.product_id + ".JP2"

    @property
    def jp2_path(self):
        prefix = "RDR/"
        postfix = ""
        if self.kind not in ["RED", "COLOR"]:
            prefix += "EXTRAS/"
        if self.kind in ["IRB"]:
            postfix = ".NOMAP"
        return prefix + self.storage_stem + postfix + ".JP2"

    @property
    def nomap_jp2_path(self):
        if self.kind in ["RED", "IRB", "RGB"]:
            return f"EXTRAS/RDR/{self.storage_stem}.NOMAP.JP2"
        else:
            raise ValueError(f"No NOMAP exists for {self.kind}.")

    @property
    def quicklook_path(self):
        if self.kind in ["COLOR", "RED"]:
            return Path("EXTRAS/RDR/") / (self.storage_stem + ".QLOOK.JP2")
        else:
            raise ValueError(f"No quicklook exists for {self.kind} products.")

    @property
    def abrowse_path(self):
        if self.kind in ["COLOR", "MIRB", "MRGB", "RED"]:
            return Path("EXTRAS/RDR/") / (self.storage_stem + ".abrowse.jpg")
        else:
            raise ValueError(f"No abrowse exists for {self.kind}")

    @property
    def browse_path(self):
        inset = ""
        if self.kind in ["IRB", "RGB"]:
            inset = ".NOMAP"
        if self.kind not in ["COLOR", "MIRB", "MRGB", "RED", "IRB", "RGB"]:
            raise ValueError(f"No browse exists for {self.kind}")
        else:
            return Path("EXTRAS/RDR/") / (self.storage_stem + inset + ".browse.jpg")

    @property
    def thumbnail_path(self):
        if self.kind in ["BG", "IR"]:
            raise ValueError(f"No thumbnail exists for {self.kind}")
        inset = ""
        if self.kind in ["IRB", "RGB"]:
            inset = ".NOMAP"
        return Path("EXTRAS/RDR/") / (self.storage_stem + inset + ".thumb.jpg")

    @property
    def nomap_thumbnail_path(self):
        if self.kind in ["RED", "IRB", "RGB"]:
            return Path("EXTRAS/RDR") / (self.storage_stem + ".NOMAP.thumb.jpg")
        else:
            raise ValueError(f"No NOMAP thumbnail exists for {self.kind} images.")

    @property
    def nomap_browse_path(self):
        if self.kind in ["RED", "IRB", "RGB"]:
            return Path("EXTRAS/RDR") / (self.storage_stem + ".NOMAP.browse.jpg")
        else:
            raise ValueError(f"No NOMAP browse exists for {self.kind} images.")

    @property
    def edr_storage_stem(self):
        return Path("EDR") / self.storage_stem

    @property
    def homepage(self) -> str:  # URL to the product's homepage
        return f"https://uahirise.org/{self.obsid}"

    def go_to_homepage(self):
        webbrowser.open(self.homepage)

# %% ../notebooks/api/04_hirise.ipynb 33
class COLOR_PRODUCT:
    def __init__(self, obsid):
        self.obsid = obsid
        # this should be reset by the subclass
        self.pathfinder = ProductPathfinder(self.obsid + "_COLOR")

    @property
    def product_id(self):
        return self.pathfinder.product_id

    @property
    def meta(self):
        color_id = self.obsid + "_COLOR"
        s = rdrindex.query("PRODUCT_ID == @color_id").squeeze()
        s.index = s.index.str.lower()
        return s

    @property
    def url(self):
        # set `name` in the subclass
        return URL(getattr(self.pathfinder, self.name + "_url"))

    @property
    def remote_path(self):
        # set `name` in subclass
        return Path(getattr(self.pathfinder, self.name + "_path"))

    @property
    def local_path(self):
        path = self.remote_path
        return storage_root / (f"{path.parent.name}/{path.name}")

    def download(self, overwrite=False):
        self.local_path.parent.mkdir(parents=True, exist_ok=True)
        if self.local_path.exists() and not overwrite:
            print("File exists. Use `overwrite=True` to download fresh.")
            return
        url_retrieve(self.url, self.local_path)

    def read(self):
        self.da = rxr.open_rasterio(self.local_path, chunks=(1, 2024, 2024))
        return self.da

    def show(self, **kwargs):
        self.read()
        return self.plot_da(**kwargs)

    def plot_da(self, xslice=None, yslice=None):
        if xslice is not None or yslice is not None:
            data = self.da.isel(x=xslice, y=yslice)
        else:
            data = self.da

        return data.hvplot.image(
            x="x",
            y="y",
            rasterize=True,
            widget_location="top_left",
            cmap="gray",
            frame_height=800,
            frame_width=800,
            flip_yaxis=True,
        )

# %% ../notebooks/api/04_hirise.ipynb 34
class RGB_NOMAP(COLOR_PRODUCT):
    def __init__(self, obsid):
        super().__init__(obsid)
        self.name = "nomap_jp2"
        self.pathfinder = ProductPathfinder(obsid + "_RGB")

# %% ../notebooks/api/04_hirise.ipynb 45
class RGB_NOMAPCollection:
    """Class to deal with a set of RGB_NOMAP products."""

    def __init__(self, obsids):
        self.obsids = obsids

    def get_urls(self):
        """Get URLs for list of obsids.

        Returns
        -------
        List[yarl.URL]
            List of URL objects with the respective PDS URL for download.
        """
        urls = []
        for obsid in self.obsids:
            rgb = RGB_NOMAP(obsid)
            urls.append(rgb.url)
        self.urls = urls
        return urls

    @property
    def local_paths(self):
        paths = []
        for obsid in self.obsids:
            rgb = RGB_NOMAP(obsid)
            paths.append(rgb.local_path)
        return paths

    def download_collection(self, **kwargs):
        lazys = []
        for obsid in self.obsids:
            rgb = RGB_NOMAP(obsid)
            lazys.append(delayed(rgb.download)(**kwargs))
        print("Launching parallel download...")
        compute(*lazys)
        print("Done.")

# %% ../notebooks/api/04_hirise.ipynb 46
class SOURCE_PRODUCT:
    """Manage SOURCE_PRODUCT id.

    Example
    -------
    'PSP_003092_0985_RED4_0'
    """

    red_ccds = ["RED" + str(i) for i in range(10)]
    ir_ccds = ["IR10", "IR11"]
    bg_ccds = ["BG12", "BG13"]
    ccds = red_ccds + ir_ccds + bg_ccds

    def __init__(self, spid, saveroot=None, check_url=True):
        tokens = spid.split("_")
        obsid = "_".join(tokens[:3])
        ccd = tokens[3]
        color, ccdno = self._parse_ccd(ccd)
        self.pid = ProductPathfinder("_".join([obsid, color]))
        self.ccd = ccd
        self.channel = tokens[4]
        self.saveroot = storage_root if saveroot is None else saveroot

    def __getattr__(self, value):
        return getattr(self.pid, value)

    def _parse_ccd(self, value):
        sep = 2 if value[:2] in ProductPathfinder.kinds else 3
        return value[:sep], value[sep:]

    @property
    def channel(self):
        return self._channel

    @channel.setter
    def channel(self, value):
        if int(value) not in [0, 1]:
            raise ValueError("channel must be in [0, 1]")
        self._channel = value

    @property
    def ccd(self):
        return self._ccd

    @ccd.setter
    def ccd(self, value):
        if value not in self.ccds:
            raise ValueError("CCD value must be in {}.".format(self.ccds))
        self._ccd = value
        if self.pid is not None:
            self.pid.color = self.color

    @property
    def color(self):
        return self._parse_ccd(self.ccd)[0]

    @property
    def ccdno(self):
        offset = len(self.color)
        return self.ccd[offset:]

    def __str__(self):
        return "{}: {}{}_{}".format(
            self.__class__.__name__, self.pid, self.ccdno, self.channel
        )

    def __repr__(self):
        return self.__str__()

    @property
    def spid(self):
        return f"{self.pid}{self.ccdno}_{self.channel}"

    @property
    def fname(self):
        return self.spid + ".IMG"

    @property
    def remote_path(self):
        return Path(self.pid.edr_storage_stem).parent / self.fname

    @property
    def url(self):
        u = baseurl / str(self.remote_path)
        if self.check_url:
            if not check_url_exists(u):
                warnings.warn(f"{u} does not exist on the server.")
        return u

    @property
    def local_path(self):
        savepath = self.saveroot / str(self.obsid) / self.fname
        return savepath

    @property
    def local_cube(self):
        return self.local_path.with_suffix(".cub")

    @property
    def stitched_cube_name(self):
        return f"{self.pid.obsid.id}_{self.ccd}.cub"

    @property
    def stitched_cube_path(self):
        return self.local_cube.with_name(self.stitched_cube_name)

    def download(self, overwrite=False):
        self.local_path.parent.mkdir(parents=True, exist_ok=True)
        if self.local_path.exists() and not overwrite:
            print("File exists. Use `overwrite=True` to download fresh.")
            return
        url_retrieve(self.url, self.local_path)

# %% ../notebooks/api/04_hirise.ipynb 61
class RED_PRODUCT(SOURCE_PRODUCT):
    "This exists to support creating a RED_PRODUCT_ID from parts of a SOURCE_PRODUCT id."

    def __init__(self, obsid, ccdno, channel, **kwargs):
        self.ccds = self.red_ccds
        super().__init__(f"{obsid}_RED{ccdno}_{channel}", **kwargs)

# %% ../notebooks/api/04_hirise.ipynb 66
class IR_PRODUCT(SOURCE_PRODUCT):
    def __init__(self, obsid, ccdno, channel):
        self.ccds = self.ir_ccds
        super().__init__(f"{obsid}_IR{ccdno}_{channel}", **kwargs)


class BG_PRODUCT(SOURCE_PRODUCT):
    def __init__(self, obsid, ccdno, channel):
        self.ccds = self.ir_ccds
        super().__init__(f"{obsid}_BG{ccdno}_{channel}", **kwargs)
